common/src/IO/MapReader.cpp

```diff
@@ -352,6 +355,30 @@ namespace TrenchBroom {
             }
         }
 
+        void MapReader::resolveBrushes(ParserStatus& status) {
+            // Temporary vector type to store the results of creating the Brush objects in parallel.
+            // The std::optional wrapper is just so we can default-initialize the vector to the
+            // right size.
+            using BrushResults = std::vector<std::optional<kdl::result<Model::Brush, Model::BrushError>>>;
+
+            BrushResults brushResults;
+            brushResults.resize(m_brushInfos.size());
+
+            // In parallel, create Brush objects (moving faces out of m_brushInfos)
+            kdl::parallel_for(m_brushInfos.size(), [&](const size_t i) {
+                BrushInfo& brushInfo = m_brushInfos[i];
+                brushResults[i] = std::make_optional(Model::Brush::create(m_worldBounds, std::move(brushInfo.faces)));
```

Comment:
Why don't you create a brush node here and store a pair of the parent and the new brush node in the vector? Then you only need to add it to the parent in the following loop and you don't need to access the brush infos anymore.

Reply:
Mostly because of logging - we log errors on brush creation failure. I didn't want to deal with trying to log from a background thread, having to add mutexes, having the logging order be nondeterministic, etc.

Reply:
You could store a `std::pair<Model::Node*, std::variant<std::unique_ptr<Model::BrushNode>, BrushError>>`. I know it's a mouthful, but I like that because it's more of a transformation pipeline which does not need to access the results or inputs of previous stages (no need to access brush infos after they were processed). But it's not terribly important.

Reply:
Ah, nice, I didn't see that possibility. Wil give it a shot later.

> because it's more of a transformation pipeline which does not need to access the results or inputs of previous stages (no need to access brush infos after they were processed). 

Yeah, that would be ideal - the current design with the `kdl::parallel_for` was a compromise to make everything work and reuse the existing `createBrush`,but what you suggested should be better.